src/library/scala/collection/Iterator.scala

```diff
@@ -470,13 +479,22 @@ trait Iterator[+A] extends TraversableOnce[A] {
   *  @note     Reuse: $consumesAndProducesIterator
   */
   @migration("`collect` has changed. The previous behavior can be reproduced with `toSeq`.", "2.8.0")
-  def collect[B](pf: PartialFunction[A, B]): Iterator[B] = {
-    val self = buffered
-    new AbstractIterator[B] {
-      private def skip() = while (self.hasNext && !pf.isDefinedAt(self.head)) self.next()
-      def hasNext = { skip(); self.hasNext }
-      def next() = { skip(); pf(self.next()) }
+  def collect[B](pf: PartialFunction[A, B]): Iterator[B] = new AbstractIterator[B] {
+    // Manually buffer to avoid extra layer of wrapping with buffered
+    private var hd: A = _
+    // Magic value indicating status: 0 means find next valid; 1 means found; -1 means done
+    private var hdStatus = 0
+    def hasNext = {
+      while (hdStatus == 0) {
+        if (self.hasNext) {
+          hd = self.next()
+          if (pf.isDefinedAt(hd)) hdStatus = 1
+        }
+        else hdStatus = -1
+      }
+      hdStatus > 0
```

Comment:
I did mean `> 0` because `hasNext`'s job is to find the next thing, and if it's not found, there isn't a `next`.  I think I had a fourth status code originally, which is why it was `> 0`; I'll fix it.  I will check to make sure the compiler and/or JVM manages to turn `final val` into inlined constants.